### ISSUE TITLE
Save website content for offline reading

### ISSUE THREAD

**Issue by Mati20187** [sissbruecker/linkding#286]
Hi, I know this may be a long shot and I'm not sure if that's within the scope of this project but I'd love to see a "reader mode" functionality implemented where the website's content is saved for offline reading (offline - independent of the bookmarked website or the WayBackMachine).

 Something like [this](https://raw.githubusercontent.com/go-shiori/shiori/master/docs/readme/comparison.png) in Shiori. I think that linkding's UI is better and that it's generally snappier than shiori, that's why I'd like to see this feature here.

**Comment by sissbruecker** [comment]
Just to clarify, this issue would be about saving a minimal, distraction-free version of the website for the purpose of making the content easy to read, and being able to access that content if the website is not accessible (being offline or website is down).

It's not about preservation or archival of the website in it's original form.

For now it's unlikely that I'll work on this, it's a lot of effort to create and maintain this feature. A 3rd-party integration could be more reasonable.

**Comment by brettinternet** [comment]
Shiori rewrote Mozilla's library [Readability](https://github.com/mozilla/readability) in Go ([go-readability](https://github.com/go-shiori/go-readability)), but the Mozilla lib is very mature as it is. If we can get python to run a JS script on the server, this could be an easy feature to implement. I really like the idea of preserving the site's content (but as you say, not the original form).

**Comment by schlamar** [comment]
There is Python implementation https://github.com/buriy/python-readability

**Comment by gitcrook** [comment]
Would really appreciate an option to save and locally archive a website independent of the Wayback Machine or any other service. Many websites or webpages make changes or are taken down constantly nowadays. I am hosting linkding as a docker in my NAS so this would be absolutely perfect. Would make linkding which is an A for me and bring it into S tier.

**Comment by gitcrook** [comment]
Forgot to add that it doesn't need to be what you were describing in terms like the ease of readability, or distraction free mode. I'd be happy with the option to screenshot and save the webpage's state with date and time. I have yet to try linkwarden but I have looked at their project and I think something similar would be nice for linkding. I am not sure if that would fall under a different "enhancement" request.

**Comment by sissbruecker** [comment]
I started working on a prototype for this. I'm currently using singlefile to generate a single HTML file from a web page, containing all CSS, assets, etc. As a bonus, the browser that captures the page is using the uBlock Origin extension to remove ads and cookie banners. There is currently just a link in the UI to view the captured HTML. In a second step there could be a view that runs Mozilla readability over the HTML to generate a reader-style view of it.

No promises yet that I'll release it, and if I do, it's going to be an alternative Docker image as it requires a ton of additional dependencies. Since the capturing is running on the server there would also be limitations such as missing logins etc.

**Comment by gitcrook** [comment]
Sounds promising, I almost forgot about uBlock origin since I don't even use any browser without installing it first. I wouldn't mind if it is an alternative docker image. Like I said would be fine if it starts or ends as lightweight as possible and just makes a quick pdf capture, maybe readable (if that requires a lot more resources). I wouldn't necessary use it for comfort offline reading but instead in this case it would be a killer feature for a quick archive of index or recommendation websites where they may change or take down their webpage for a new one. But I could see a lot of other uses since I know people like to archive posts or tweets before they are taken down.

**Comment by gitcrook** [comment]
After this discussion I looked for self hosted website archive solutions and found archive box. I really like it and will continue to use it but it’s overkill for most of the time. I also found the web browser extension Single File from: https://github.com/gildas-lormeau/SingleFile and it was basically was I was looking for. Wanted to come back to this and quickly ask if you had considered that instead of adding the function to save an html offline copy of the website, making an extensions similar to those we  already have but something like a fork of single file that saves to a folder indicated by linkding docker and which its interface can quickly lookup based on the website? Either a light implementation into the docker of saving HTML copy or this would be fine in my opinion.
